Задачи по Pandas

V 0.8 14.09.2020

Датасеты:
* sp_data2.csv 
* sp500hst.txt 
* kings.csv 
* titanic.csv 
* e-commerce/*
* avocado_v1
* us-state-population
* us-state-shortcode

In [160]:
import pandas as pd
import numpy as np

---

## 1. Введение

### Датасеты:
* sp_data2.csv
* sp500hst.txt

1.1 Задание: загрузить данные из файлов (приложены к письму):
* excel-comp-data.xlsx
* sp500hst.txt 

In [84]:
dt1 = pd.read_csv('sp_data2.csv', sep=';', header=None)
dt2 = pd.read_csv('sp500hst.txt', header=None)
print(dt1, dt2)

         0                1     2
0     AAPL            Apple  3.6%
1     AMZN       Amazon.com  3.2%
2    GOOGL         Alphabet  3.1%
3     GOOG         Alphabet  3.1%
4     MSFT        Microsoft  3.0%
..     ...              ...   ...
500    SCG            SCANA  0.0%
501    AIZ         Assurant  0.0%
502    AYI    Acuity Brands  0.0%
503    HRB        H&R Block  0.0%
504    RRC  Range Resources  0.0%

[505 rows x 3 columns]                0    1      2        3       4      5      6
0       20090821    A  25.60  25.6100  25.220  25.55  34758
1       20090824    A  25.64  25.7400  25.330  25.50  22247
2       20090825    A  25.50  25.7000  25.225  25.34  30891
3       20090826    A  25.32  25.6425  25.145  25.48  33334
4       20090827    A  25.50  25.5700  25.230  25.54  70176
...          ...  ...    ...      ...     ...    ...    ...
122569  20100813  ZMH  51.72  51.9000  51.380  51.44  14561
122570  20100816  ZMH  51.13  51.4700  50.600  51.00  13489
122571  20100817  ZMH  51.14

1.2 Загрузить данные из файла sp500hst.txt, обозначить столбцы (в соответствии с содержимым).
    * "date", "ticker", "open", "high", "low", "close", "volume"

In [85]:
dt2.columns=["date", "ticker", "open", "high", "low", "close", "volume"]
dt2

date ticker   open     high     low  close  volume
0       20090821      A  25.60  25.6100  25.220  25.55   34758
1       20090824      A  25.64  25.7400  25.330  25.50   22247
2       20090825      A  25.50  25.7000  25.225  25.34   30891
3       20090826      A  25.32  25.6425  25.145  25.48   33334
4       20090827      A  25.50  25.5700  25.230  25.54   70176
...          ...    ...    ...      ...     ...    ...     ...
122569  20100813    ZMH  51.72  51.9000  51.380  51.44   14561
122570  20100816    ZMH  51.13  51.4700  50.600  51.00   13489
122571  20100817    ZMH  51.14  51.6000  50.890  51.21   20498
122572  20100819    ZMH  51.63  51.6300  50.170  50.22   18259
122573  20100820    ZMH  50.03  50.5500  49.480  49.82   17792

[122574 rows x 7 columns]

1.3

1.3.1 Создать столбец со значеними средними, между значением дневного минимума и максимума.

In [86]:
dt2['avg_high_low'] = (dt2["high"]+dt2["low"])/2
dt2

date ticker   open     high     low  close  volume  avg_high_low
0       20090821      A  25.60  25.6100  25.220  25.55   34758      25.41500
1       20090824      A  25.64  25.7400  25.330  25.50   22247      25.53500
2       20090825      A  25.50  25.7000  25.225  25.34   30891      25.46250
3       20090826      A  25.32  25.6425  25.145  25.48   33334      25.39375
4       20090827      A  25.50  25.5700  25.230  25.54   70176      25.40000
...          ...    ...    ...      ...     ...    ...     ...           ...
122569  20100813    ZMH  51.72  51.9000  51.380  51.44   14561      51.64000
122570  20100816    ZMH  51.13  51.4700  50.600  51.00   13489      51.03500
122571  20100817    ZMH  51.14  51.6000  50.890  51.21   20498      51.24500
122572  20100819    ZMH  51.63  51.6300  50.170  50.22   18259      50.90000
122573  20100820    ZMH  50.03  50.5500  49.480  49.82   17792      50.01500

[122574 rows x 8 columns]

1.3.2 Рассчитать среднее значение показателей для каждого из столбцов 3-6.

In [24]:
names = dt2.columns[2:6].to_list()
for i in names:
    print('mean of the column: ', i,'  ', np.mean(dt2[i]))

mean of the column:  open    42.5954576590471
mean of the column:  high    43.10224338767034
mean of the column:  low    42.054463665213184
mean of the column:  close    42.60186484817329


1.4

1.4.1 Создать DataFrame состоящий из строк относящихся только к 13 августа 2010 года. 

In [87]:
dt2['date'] = pd.to_datetime(dt2['date'], format='%Y%m%d')
dt_141 = dt2[dt2['date'] == '2010-08-10']
dt_141

date ticker    open     high     low   close  volume  \
237    2010-08-10      A   29.44   29.680   29.13   29.46   34866   
482    2010-08-10     AA   11.43   11.490   11.18   11.35  264545   
723    2010-08-10   AAPL  259.85  260.450  257.55  259.41  161399   
968    2010-08-10    ABC   29.82   30.085   29.65   29.97   28542   
1213   2010-08-10    ABT   50.67   51.700   50.59   51.47   89666   
...           ...    ...     ...      ...     ...     ...     ...   
121382 2010-08-10    XRX    9.66    9.720    9.38    9.62  147205   
121832 2010-08-10   YHOO   14.34   14.460   14.20   14.35   96579   
122077 2010-08-10    YUM   42.43   42.720   42.00   42.39   33834   
122321 2010-08-10   ZION   21.43   21.720   21.07   21.46   32541   
122566 2010-08-10    ZMH   53.61   54.400   53.29   53.90   21266   

        avg_high_low  
237          29.4050  
482          11.3350  
723         259.0000  
968          29.8675  
1213         51.1450  
...              ...  
121382        9.5500  
121832       14.3300  
122077       42.3600  
122321       21.3950  
122566       53.8450  

[498 rows x 8 columns]

1.4.1.1 Из DataFrame, созданного в 1.4.1, cоздать серию из 30 первых имен 

In [42]:
dt_141.iloc[:30]['ticker']

237        A
482       AA
723     AAPL
968      ABC
1213     ABT
1270     ACE
1623    ADBE
1868     ADI
2113     ADM
2356     ADP
2598    ADSK
2843     AEE
3088     AEP
3333     AES
3578     AET
3823     AFL
4068     AGN
4313     AIG
4558     AIV
4803     AIZ
5046    AKAM
5291     AKS
5536     ALL
5779    ALTR
6022    AMAT
6267     AMD
6511    AMGN
6755     AMP
7000     AMT
7243    AMZN
Name: ticker, dtype: object

1.4.2 Из DataFrame, созданного в 1.4.1, создать новый DataFrame, индексом в котором являются имена тикеров.

In [45]:
dt_142 = dt_141.set_index('ticker')
dt_142

date    open     high     low   close  volume       avg  20100813
ticker                                                                        
A      2010-08-10   29.44   29.680   29.13   29.46   34866   29.4050  20100813
AA     2010-08-10   11.43   11.490   11.18   11.35  264545   11.3350  20100813
AAPL   2010-08-10  259.85  260.450  257.55  259.41  161399  259.0000  20100813
ABC    2010-08-10   29.82   30.085   29.65   29.97   28542   29.8675  20100813
ABT    2010-08-10   50.67   51.700   50.59   51.47   89666   51.1450  20100813
...           ...     ...      ...     ...     ...     ...       ...       ...
XRX    2010-08-10    9.66    9.720    9.38    9.62  147205    9.5500  20100813
YHOO   2010-08-10   14.34   14.460   14.20   14.35   96579   14.3300  20100813
YUM    2010-08-10   42.43   42.720   42.00   42.39   33834   42.3600  20100813
ZION   2010-08-10   21.43   21.720   21.07   21.46   32541   21.3950  20100813
ZMH    2010-08-10   53.61   54.400   53.29   53.90   21266   53.8450  20100813

[498 rows x 8 columns]

1.4.3 Создать DataFrame состоящий из строк относящихся только тикеру ALTR.

In [58]:
dt_143 = dt2[dt2['ticker'] == 'ALTR']
dt_143

date ticker   open   high    low  close  volume  avg_high_low
5544 2009-08-21   ALTR  19.36  19.50  19.15  19.48   54580        19.325
5545 2009-08-24   ALTR  19.51  19.72  19.04  19.13   42059        19.380
5546 2009-08-25   ALTR  19.13  19.39  18.95  19.05   48539        19.170
5547 2009-08-26   ALTR  19.11  19.41  18.99  19.28   42058        19.200
5548 2009-08-27   ALTR  19.23  19.32  18.88  19.08   63271        19.100
...         ...    ...    ...    ...    ...    ...     ...           ...
5782 2010-08-13   ALTR  25.30  25.50  24.94  24.94   69416        25.220
5783 2010-08-16   ALTR  24.82  25.31  24.53  25.11   60276        24.920
5784 2010-08-17   ALTR  25.22  25.88  25.04  25.54   89329        25.460
5785 2010-08-19   ALTR  25.61  25.96  25.24  25.50   61815        25.600
5786 2010-08-20   ALTR  25.44  25.83  25.32  25.60   54383        25.575

[243 rows x 8 columns]

1.5

1.5.1 Создать DataFrame состоящий из строк относящихся только тикеру WAT и объединить его с DataFrame для ALTR. 

1.5.2 Создать DataFrame в котором присутствует столбец, отражающий разницу в объемах торгов по WAT и ALTR в одинаковые дни и содержит исходные данны об объеме трогов этими акциями.

1.6 Выяввиь столбцы содержащие минимальные и максимальные значения (если такие столбцы есть). (Проверить, есть ли среди столбцов, столбец с минимальным значением курса акций.)

1.7 По данным из 'sp500hst.txt' при помощи Pandas расчитать максимальные значения в 3-6 столбца для дат после 20100801

In [43]:
dt2[dt2['date'] > '2010-08-01'].iloc[:,2:7].max(axis=0)

open          505.850
high          508.600
low           503.559
close         508.100
volume    5741998.000
dtype: float64

1.8 По данным из 'sp500hst.txt' при помощи Pandas расчитать средний объем торгов (обеъм торгов - в последнем столбце) при значениях 3го столбца больше 50.

In [13]:
dt2[dt2['open'] > 50].iloc[:,6].mean(axis=0)

36467.08664563617

## 2.Даты

* Группа функций pandas.Series.dt.<Имя функции>
* Например, см.: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.dt.day.html

### Датасеты:
* sp_data2.csv
* sp500hst.txt

2.1 Из DataFrame, созданного в 1.4.3, создать DataFrame индексом в котором являются даты.

2.2 В DataFrame, созданном в 1.4.3, добавить столбец, содержащий только число месяца, к которому относится дата.

In [59]:
import datetime as dt
dt_143['m'] = dt_143['date'].dt.month
dt_143

D:\Anaconda\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


date ticker   open   high    low  close  volume  avg_high_low  m
5544 2009-08-21   ALTR  19.36  19.50  19.15  19.48   54580        19.325  8
5545 2009-08-24   ALTR  19.51  19.72  19.04  19.13   42059        19.380  8
5546 2009-08-25   ALTR  19.13  19.39  18.95  19.05   48539        19.170  8
5547 2009-08-26   ALTR  19.11  19.41  18.99  19.28   42058        19.200  8
5548 2009-08-27   ALTR  19.23  19.32  18.88  19.08   63271        19.100  8
...         ...    ...    ...    ...    ...    ...     ...           ... ..
5782 2010-08-13   ALTR  25.30  25.50  24.94  24.94   69416        25.220  8
5783 2010-08-16   ALTR  24.82  25.31  24.53  25.11   60276        24.920  8
5784 2010-08-17   ALTR  25.22  25.88  25.04  25.54   89329        25.460  8
5785 2010-08-19   ALTR  25.61  25.96  25.24  25.50   61815        25.600  8
5786 2010-08-20   ALTR  25.44  25.83  25.32  25.60   54383        25.575  8

[243 rows x 9 columns]

2.3 Из DataFrame выбрать только строки, относящиеся к 13 числу месяца.

2.4 Для тикера AAPL подсчитать, сколько дней прошло между максимальным и минимальным значением цены акции на закрытии рынка.

2.5 Для каждого тикера подсчитать, максимальную цену акции и определить, в какой день недели чаще всего достигается максимум цены акции.

In [174]:
tickers = np.unique(dt2['ticker']) # собираем все тикеры
dt2['weekday'] = list(map(lambda i: i.weekday(), dt2['date'])) # столбец с днем недели (в виде числа)
dt22 = dt2.set_index('ticker') # сортируем по тикеру
result = {}
for m in range(len(tickers)):
    line = dt22.loc[tickers[m]].iloc[:,7] # series с днями недели по тикеру
    line = line.reset_index()
    weekdays_counter = {0:0, 1:0, 2:0, 3:0, 4:0} # словарь-счетчик повторений максимумов
    for i in range (0, len(line['weekday'])-4):
        if line['weekday'][i]==0 and line['weekday'][i+4]==4:
            k = dt2['high'][i:i+4].argmax() # находим день недели с максимумом за неделю
            val = weekdays_counter.get(k)
            weekdays_counter.update({k:val+1}) # обновляем счетчик
    max_day = max(list(weekdays_counter.values())) # находим день недели с максимальным числом повторений максимумов по тикеру
    result[tickers[m]] = list(weekdays_counter.keys())[list(weekdays_counter.values()).index(max_day)]
result

{'A': 0,
 'AA': 0,
 'AAPL': 0,
 'ABC': 0,
 'ABT': 0,
 'ACE': 3,
 'ACS': 0,
 'ADBE': 0,
 'ADI': 0,
 'ADM': 0,
 'ADP': 0,
 'ADSK': 0,
 'AEE': 0,
 'AEP': 0,
 'AES': 0,
 'AET': 0,
 'AFL': 0,
 'AGN': 0,
 'AIG': 0,
 'AIV': 0,
 'AIZ': 0,
 'AKAM': 0,
 'AKS': 0,
 'ALL': 0,
 'ALTR': 0,
 'AMAT': 0,
 'AMD': 0,
 'AMGN': 0,
 'AMP': 0,
 'AMT': 0,
 'AMZN': 0,
 'AN': 0,
 'ANF': 0,
 'AOC': 0,
 'AON': 0,
 'APA': 0,
 'APC': 0,
 'APD': 0,
 'APH': 0,
 'APOL': 0,
 'ARG': 3,
 'ATI': 0,
 'AVB': 0,
 'AVP': 0,
 'AVY': 0,
 'AXP': 0,
 'AYE': 0,
 'AZO': 0,
 'BA': 0,
 'BAC': 0,
 'BAX': 0,
 'BBBY': 0,
 'BBT': 0,
 'BBY': 0,
 'BCR': 0,
 'BDK': 0,
 'BDX': 0,
 'BEN': 0,
 'BF.B': 0,
 'BHI': 0,
 'BIG': 0,
 'BIIB': 0,
 'BJS': 0,
 'BK': 0,
 'BLL': 0,
 'BMC': 0,
 'BMS': 0,
 'BMY': 0,
 'BNI': 0,
 'BRCM': 0,
 'BRK.B': 2,
 'BSX': 0,
 'BTU': 0,
 'BXP': 0,
 'C': 0,
 'CA': 0,
 'CAG': 0,
 'CAH': 0,
 'CAM': 0,
 'CAT': 0,
 'CB': 0,
 'CBE': 0,
 'CBG': 0,
 'CBS': 0,
 'CCE': 0,
 'CCL': 0,
 'CEG': 0,
 'CELG': 0,
 'CEPH': 0,
 'CERN': 2,
 '

## 3. Работа со строковыми значениями в Pandas

* Группа функций pandas.Series.str.<Имя функции>
* Например, см.: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.capitalize.html

### Датасеты:
* kings.csv

In [179]:
k = pd.read_csv('kings_v2.csv', sep=';')
k

N                                               King
0    1           Карл II Лысый (843—877), император с 875
1    2                         Людовик II Заика (877—879)
2    3                              Людовик III (879—882)
3    4    Карломан II (879—884), соправитель Людовика III
4    5  Карл III Толстый (884—888), король Восточно-Фр...
5    6  Эд I (888—898), из династии Робертинов, граф П...
6    7   Гвидо Сполетский (888—888), титулярный король...
7    8                     Карл III Простоватый (898—922)
8    9         Роберт I (922—923), из династии Робертинов
9   10           Рауль I (923—936), из династии Бозонидов
10  11                               Людовик IV (936—954)
11  12                                   Лотарь (954—986)
12  13    Карл Нижнелотарингский (978—978), коронован ...
13  14                        Людовик V Ленивый (986—987)

3.1 Загрузить датасет kings.csv в датафрэйм и добавить в него столбец Name, в котором содержатся имена королей (без порядковых номеров) 

In [180]:
k['Name'] = k['King'].apply(lambda x: str(x).split('(')[0].strip())
k

N                                               King  \
0    1           Карл II Лысый (843—877), император с 875   
1    2                         Людовик II Заика (877—879)   
2    3                              Людовик III (879—882)   
3    4    Карломан II (879—884), соправитель Людовика III   
4    5  Карл III Толстый (884—888), король Восточно-Фр...   
5    6  Эд I (888—898), из династии Робертинов, граф П...   
6    7   Гвидо Сполетский (888—888), титулярный король...   
7    8                     Карл III Простоватый (898—922)   
8    9         Роберт I (922—923), из династии Робертинов   
9   10           Рауль I (923—936), из династии Бозонидов   
10  11                               Людовик IV (936—954)   
11  12                                   Лотарь (954—986)   
12  13    Карл Нижнелотарингский (978—978), коронован ...   
13  14                        Людовик V Ленивый (986—987)   

                      Name  
0            Карл II Лысый  
1         Людовик II Заика  
2              Людовик III  
3              Карломан II  
4         Карл III Толстый  
5                     Эд I  
6         Гвидо Сполетский  
7     Карл III Простоватый  
8                 Роберт I  
9                  Рауль I  
10              Людовик IV  
11                  Лотарь  
12  Карл Нижнелотарингский  
13       Людовик V Ленивый

In [181]:
print(k)

     N                                               King  \
0    1           Карл II Лысый (843—877), император с 875   
1    2                         Людовик II Заика (877—879)   
2    3                              Людовик III (879—882)   
3    4    Карломан II (879—884), соправитель Людовика III   
4    5  Карл III Толстый (884—888), король Восточно-Фр...   
5    6  Эд I (888—898), из династии Робертинов, граф П...   
6    7   Гвидо Сполетский (888—888), титулярный король...   
7    8                     Карл III Простоватый (898—922)   
8    9         Роберт I (922—923), из династии Робертинов   
9   10           Рауль I (923—936), из династии Бозонидов   
10  11                               Людовик IV (936—954)   
11  12                                   Лотарь (954—986)   
12  13    Карл Нижнелотарингский (978—978), коронован ...   
13  14                        Людовик V Ленивый (986—987)   

                      Name  
0            Карл II Лысый  
1         Людовик II Заика

3.2 Для датафрэйма kings загрузить порядоковые нормера королей и отфильторвать строки соответствующие королям с номерами "I".

3.3 Для датафрэйма kings, для столбца King получить значения максимума и минимума для чисел, являющихся последним токеном в строке.

In [211]:
cuts = list(map(lambda q: q[q.rfind('—')+1:q.rfind('—')+4], k['King']))
min_ = 1000
max_ = 0
for i in cuts:
    try:
        i = int(i)
        if i<min_:
            min_=i
        if i>max_:
            max_=i
    except:
        None
min_, max_

(877, 987)

In [209]:
a = 'iii'
int(a)

ValueError: invalid literal for int() with base 10: 'iii'

3.4* Для датафрэйма kings, для столбца King с помощью регулярного выражения извелечь значения начала и конца правления королей, записанные в формате (год-год).

## 4. Работа с пропущенными значениями

### Датасеты:
* titanic.csv

In [236]:
tit = pd.read_csv('titanic.csv', sep=',')
tit

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

4.1 Подсчитать количество пропущенных значений для всех столбцов датасета titanic.csv

In [232]:
list(map(lambda i: tit.loc[:,i].isna().sum(), tit))

[0, 0, 0, 0, 0, 177, 0, 0, 0, 0, 687, 2]

In [233]:
# проверка
tit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


4.2 Удалить из датафрейма полученного из titanic.csv все строки, в которых есть хотя бы один пропуск.

In [237]:
tit_na = tit.dropna(how = 'any')
tit.shape, tit_na.shape

((891, 12), (183, 12))

4.3 Удалить из датафрейма полученного из titanic.csv все строки, в которых есть два и более пропуска.

4.4 В датафрейме, полученном из titanic.csv, заменить все пропущенные числовые значения на значения равне среднему от аналогичного значения в предыдущей и последующей строках.

4.5 В датафрейме, полученном из titanic.csv, заменить все пропущенные числовые значения возраста на значения равне среднему значению для представителей этого пола (не выполнять операцию, если неизвестен и пол и возраст).

## 5.Сводные таблицы в Pandas
* см.: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html

### Датасеты:
* titanic.csv

5.1 С помощью сводных таблиц определить какая доля выживших в разных классах для мужчин и женщин была на Титанике (датасет titanic.csv)

5.2 С использованием сводных таблиц подсчитать количество мужчин и женщин плывших в разных классов на Титанике (датасет titanic.csv)

5.3 С использованием сводных таблиц определить какая доля выживших была на Титанике для мужчин и женщин, разных классов и разных категорий возрастов (Категории возрастов: (0, 10), (10, 30), (30, 60), (60, 80)) (датасет titanic.csv)

## 6.Объединение наборов данных

### Датасеты:
* sp_data2.csv
* sp500hst.txt
* us-state-population
* us-state-shortcode
* e-commerce/*
* avocado_v1

6.1 Добавить столбец с расшифровкой названия тикера используя данные из файла sp_data2.csv . В случае нехватки данных об именах тикеров корректно обработать их.

6.2 Вывести тикеры, для которых не были найдены рсшифровки. 

6.3 Отранжировать штаты по численности населения меньше 18 лет в 1990 году. В качестве индекса таблицы использовать название штата

6.4 Объединить данные о торговле из /e-commerce в один датафрейм, добавив столбец 'Country', в котором указано название страны

6.5

6.5.1. Загрузить датасет avocado. 

6.5.2. Преобразовать столбец Date в формате datetime

6.5.3. При помощи метода join создать новый датафрейм, в котором столбцами являются скользящие средние цен за 10 дней, а индексом - даты

6.5.4. Изобразить на одной фигуре графики полученных временных рядом

## 7. GroupBy

In [ ]:
### Датасеты:
* sp_data2.csv
* sp500hst.txt
* us-state-population
* us-state-shortcode
* e-commerce/*
* avocado_v1

7.1 Рассчитать суммарный объем торгов (столбец 7) для для одинаковых значений (тикеров) из столбца 2 

7.2 Для каждого тикера подсчитать, сколько раз он встречается.

7.3 Расчитать среднее значение показателей для каждого из столбцов 3-6 для одинаковых значений (тикеров) из столбца 2 и сохранить в новом DataFrame

7.4 Добавить в DataFrame новый столбец, содержащий объем торгов (столбец 7), нормированный по среднему значению объема торгов для данного тикера.

7.5 Для каждой из стран в датафрейме, полученном в 6.4 найти 3 самых дорогих проданных товара

7.6 Разбить датафрейм, полученный в 6.4, на 4 файла, в зависимости от принадлежности значения столбца 'UnitPrice' интервалам:

1) $ UnitPrice \in (q_0, q_1] $

2) $ UnitPrice \in (q_1, q_2] $

3) $ UnitPrice \in (q_2, q_3] $

4) $ UnitPrice \in (q_3, q_4] $

7.7 Создать новый столбец UnitPriceWithMean, значения которого совпадают с UnitPrice, если UnitPrice >= 2 и равны среднему значению цен товаров, больших двух, в этой стране, в противном случае

7.8 Для каждого описания товара найти список стран, в которых этот товар был продан. Описание товара считаем корректным, если в нем все символы строки, поддерживающие приведение к регистру, приведены к верхнем